flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java

```diff
@@ -0,0 +1,830 @@
+                    return matched;
+                }
+
+                // The association count for the *current* depth is reset before recursing.
+                // This is because `associations[depth]` should reflect matches found for the
+                // current
+                // combinations of records in `currentRows` with the matches it finds to the right.
+                associations[depth] = 0;
+            }
+
+            // During `CALCULATE_MATCHES` at the `inputId` level, the primary goal is to
+            // determine associations for the input record's level. Recursion to deeper
+            // levels (joins to the right) is not needed for this specific count at this stage,
+            // as the input record's participation and further joins are handled by
+            // `processInputRecord` or subsequent recursive calls in the `EMIT_RESULTS` phase.
+            if (phase == JoinPhase.CALCULATE_MATCHES && inputId == depth) {
+                continue;
+            }
+
+            recursiveMultiJoin(depth + 1, input, inputId, currentRows, associations, phase);
+        }
+
+        // Returns whether any record at this level matched the local condition.
+        return matched;
+    }
+
+    private static boolean isInnerJoin(boolean isLeftJoin) {
+        return !isLeftJoin;
+    }
+
+    private void processWithNullPadding(
+            int depth,
+            RowData input,
+            int inputId,
+            RowData[] currentRows,
+            int[] associations,
+            JoinPhase phase)
+            throws Exception {
+
+        // Recursion continues with a null row at the current depth. This means the current join
+        // emits a null padded output which is used for the next levels.
+        // By continuing the recursion, we allow those deeper conditions to be evaluated and
+        // ensures that the null padding correctly propagates to the join chain.
+        currentRows[depth] = nullRows.get(depth);
+        recursiveMultiJoin(depth + 1, input, inputId, currentRows, associations, phase);
+    }
+
+    private void processInputRecord(
+            int depth,
+            RowData input,
+            int inputId,
+            RowData[] currentRows,
+            int[] associations,
+            boolean matched)
+            throws Exception {
+
+        boolean isLeftJoin = isLeftJoinAtDepth(depth);
+        RowKind inputRowKind = input.getRowKind(); // Preserve original RowKind
+
+        // --- Left Join Retraction Handling ---
+        // For an incoming INSERT/UPDATE_AFTER on the right side of a LEFT join,
+        // if the corresponding left-side row previously had no matches (indicated by `!matched`),
+        // it would have resulted in a null-padded output. This new record might form a
+        // valid join, so the previous null-padded output must be retracted.
+        if (isUpsert(input) && isLeftJoin && !matched) {
```

Comment:
nit: I wonder if we can reuse `UpdateMode.UPSERT`, `RETRACT` here to simplify the code a bit...

Reply:
I searched the codebase for updateMode.UPSERT and found this in Plannerbase
```
val (needUpdateBefore, withChangeFlag) = outputConversion.getUpdateMode match {
          case UpdateMode.RETRACT => (true, true)
          case UpdateMode.APPEND => (false, false)
          case UpdateMode.UPSERT => (false, true)
        }
```

Can you give me a simple example on how we could use this here?